api/transport/transporttest/reqres_test.go

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transporttest
+
+import (
+	"bytes"
+	"io/ioutil"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/api/transport"
+)
+
+func TestRequestMatcher(t *testing.T) {
+	resMatcher := NewRequestMatcher(t, &transport.Request{
+		ID:              "id",
+		Host:            "host-name",
+		Environment:     "testing",
+		Caller:          "caller",
+		Service:         "service-name",
+		Transport:       "transport",
+		Encoding:        "encoding",
+		Procedure:       "procedure",
+		Headers:         transport.NewHeaders().With("foo", "bar"),
+		ShardKey:        "shardkey",
+		RoutingKey:      "routingkey",
+		RoutingDelegate: "routingdelegate",
+		Body:            ioutil.NopCloser(bytes.NewReader([]byte("my body"))),
+	})
+
+	t.Run("non-request", func(t *testing.T) {
```

Comment:
(optional)alternative: we may extract the test cases and use a for loop to execute them.

Reply:
I am also more accustomed to table-driven tests vs. subtests. With the exception of the first one, you could easily rewrite as table-driven, but subtests here are just as clear.